src/main/java/com/fasterxml/jackson/core/base/ParserMinimalBase.java

```diff
@@ -782,6 +795,22 @@ protected final void _wrapError(String msg, Throwable t) throws JsonParseExcepti
         throw _constructReadException(msg, t);
     }
 
+    protected final JsonToken _updateToken(final JsonToken token) throws StreamConstraintsException {
+        _currToken = token;
+        if (streamReadConstraints().hasMaxTokenCount() && token != JsonToken.NOT_AVAILABLE) {
```

Comment:
@cowtowncoder I'm doing the hasMaxTokenCount check here to try to minimise the impact of this code when hasMaxTokenCount=false. Testing for hasMaxTokenCount in the validateTokenCount will mean that we do a fair amount of unnecessary work if hasMaxTokenCount=false.

This change needs more testing and microbenchmarking but this is what I think is the best initial strategy.

Reply:
As a further optimization, I think we should just have

    protected final boolean _hasTokenCountLimit;

initialized on parser being constructed. But need to see how to do that given that for now `StreamReadConstraints` are not properly passed to `ParserMinimalBase`. I can probably change code to pass it tho. Will create an issue and try to get that done.

I think we can also remove `JsonToken.NOT_AVAILABLE` check: there's the other method to call and only non-blocking/Async parsers (JSON, Smile) ever set that token, nothing else.